Amplify/Categories/API/Operation/NondeterminsticOperation.swift

```diff
@@ -0,0 +1,99 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+
+import Combine
+/**
+ A non-deterministic operation offers multiple paths to accomplish its task.
+ It attempts the next path if all preceding paths have failed with an error that allows for continuation.
+ */
+enum NondeterminsticOperationError: Error {
+    case totalFailure
+    case cancelled
+}
+
+final class NondeterminsticOperation<T> {
+    /// operation that to be eval
+    typealias Operation = () async throws -> T
+    typealias OnError = (Error) -> Bool
+
+    private let operations: AsyncStream<Operation>
+    private var shouldTryNextOnError: OnError = { _ in true }
+    private var cancellables = Set<AnyCancellable>()
+    private var task: Task<Void, Never>?
+
+    deinit {
+        cancel()
+    }
+
+    init(operations: AsyncStream<Operation>, shouldTryNextOnError: OnError? = nil) {
+        self.operations = operations
+        if let shouldTryNextOnError {
+            self.shouldTryNextOnError = shouldTryNextOnError
+        }
+    }
+
+    convenience init(
+        operationStream: AnyPublisher<Operation, Never>,
+        shouldTryNextOnError: OnError? = nil
+    ) {
+        var cancellables = Set<AnyCancellable>()
+        self.init(
+            operations: AsyncStream<Operation> { continuation in
+                operationStream.sink { _ in
+                    continuation.finish()
+                } receiveValue: { operation in
+                    continuation.yield(operation)
+                }.store(in: &cancellables)
+            },
+            shouldTryNextOnError: shouldTryNextOnError
+        )
+        self.cancellables = cancellables
+    }
+
+    /// Synchronous version of executing the operations
+    func execute() -> Future<T, Error> {
+        Future { [weak self] promise in
+            self?.task = Task { [weak self] in
+                do {
+                    if let self {
+                        promise(.success(try await self.run()))
+                    } else {
+                        promise(.failure(NondeterminsticOperationError.cancelled))
+                    }
+                } catch {
+                    promise(.failure(error))
+                }
+            }
+        }
+    }
+
+    /// Asynchronous version of executing the operations
+    func run() async throws -> T {
+        for await operation in operations {
+            if Task.isCancelled {
+                throw NondeterminsticOperationError.cancelled
+            }
+            do {
+                return try await operation()
+            } catch {
+                if shouldTryNextOnError(error) {
+                    continue
+                } else {
+                    throw error
+                }
+            }
+        }
+        throw NondeterminsticOperationError.totalFailure
+    }
```

Comment:
Neat abstraction you have here. In what context would a caller choose to use `execute()` or `run()`? I can see RetryableGraphQLOperation uses this, why does it use both these methods?

Reply:
I think the main difference is `run` can be invoked in async context. While `execute` can be invoked in sync code with call back style. Also, `execute` API uses `run` internally.